draft-ietf-tls-esni.md

```diff
@@ -580,55 +594,46 @@ reuse.
-then processes the extension's "response_type" field:
-
-- If the value is "esni_accept", the client MUST check that the extension's
-  "nonce" field matches ClientESNIInner.nonce and otherwise abort the
-  connection with an "illegal_parameter" alert. The client then proceeds
-  with the connection as usual, authenticating the connection for the origin
-  server.
-
-- If the value is "esni_retry_request", the client proceeds with the handshake,
-  authenticating for ESNIKeys.public_name as described in
-  {{auth-public-name}}. If authentication or the handshake fails, the client
-  MUST return a failure to the calling application. It MUST NOT use the retry
-  keys.
-
-  Otherwise, when the handshake completes successfully with the public name
-  authenticated, the client MUST abort the connection with an "esni_required"
-  alert. It then processes the "retry_keys" field from the server's
-  "encrypted_server_name" extension.
-
-  If one of the values contains a version supported by the client, it can regard
-  the ESNI keys as securely replaced by the server. It SHOULD retry the
-  handshake with a new transport connection, using that value to encrypt the
-  SNI. The value may only be applied to the retry connection. The client
-  MUST continue to use the previously-advertised keys for subsequent
-  connections. This avoids introducing pinning concerns or a tracking vector,
-  should a malicious server present client-specific retry keys to identify
-  clients.
-
-  If none of the values provided in "retry_keys" contains a supported version,
-  the client can regard ESNI as securely disabled by the server. As below, it
-  SHOULD then retry the handshake with a new transport connection and ESNI
-  disabled.
-
-- If the field contains any other value, the client MUST abort the connection
-  with an "illegal_parameter" alert.
-
-If the server negotiates an earlier version of TLS, or if it does not
-provide an "encrypted_server_name" extension in EncryptedExtensions, the
+When the negotiated version is TLS 1.3 and client receives the first encrypted
+record, it generates two traffic secrets: one using the ClientHello before the
+ESNI transformation as the handshake transcript, and the other using the
+ClientHello post ESNI transformation. One of these two traffic secrets is
+promoted to the handshake traffic secret using the following procedure.
+
+If the encrypted record can be sucessfully decrypted using the former traffic
+secret, it means that the server has received and recognized the protected
+server name. Then, that traffic secret is promoted to the handshake traffic
+secret, and the handshake continues accordingly.
+
+Otherwise, the latter traffic secret is promoted to the handshake traffic
+secret, as the server was unable to unwind the ESNI transformation.  The client
+then proceeds with the handshake, authenticating for ESNIKeys.public_name as
+described in {{auth-public-name}}.
+
```

Comment:
Otherwise, the *post-transformation* traffic secret

Reply:
That's a very clean mechanism. I like it.

Reply:
Done in c259481.